Speed up fancy indexing

iguinn opened the pull request:
See issue https://github.com/h5py/h5py/issues/2536

Fancy indexing goes very slowly when using a large number of indices. This is because the underlying hdf5 algorithm for adding hyperslab to a selection scales as O(n^2) since each new block must be checked for overlaps with every other block. This is especially noticeable once you have thousands of elements in a fancy index array. The problem and fortunately also a solution are laid out here: https://forum.hdfgroup.org/t/quadratic-runtime-of-selecting-n-hyperslabs/12555

Instead of iterating through each element in the fancy index array and adding a hyperslab one at a time, we can use a merge algorithm, since combining two non-overlapping selections is still linear in the size of the selection. This results in  O(nlogn) complexity.

To implement this, I exposed H5Scombine_hyperslab, H5Scombine_select, H5Smodify_select, and H5Sselect_copy to the low level h5s interface (only modify_select and select_copy were needed for this speedup though). These functions were introduced in hdf5 version 1.10.6. I then modified the Selector to use a recursive algorithm to build up the selection by splitting the array in two and using combine_select to merge the resulting h5s objects. If the size of the fancy index is <16, it will iterate through the list of indices as previously (I tested a few limits here and found 16 to perform best).

![FancyIndexTiming_slow](https://github.com/user-attachments/assets/c818371e-17f1-48e7-9cc7-2a77fced5478)
![FancyIndexTiming_fast](https://github.com/user-attachments/assets/4e109507-b9d2-4c04-9fa7-c177c4c715a7)

The resulting algorithm is dramatically faster than the old one. The above test used a dataspace of size 1000000 and tested different sizes of randomly selected fancy index arrays. This test can be repeated with:
```
def test_fancy_index(n_space, n_samps):
    entries = sorted(random.sample(range(n_space), n_samps))
    h5s_src = h5py.h5s.create_simple((n_space,))
    h5s_src.select_none()
    
    t0 = time.time()
    sel = h5py._selector.Selector(h5s_src)
    sel.make_selection((entries,))
    return time.time() - t0
```

I have not yet added/modified any tests; I'm unsure of what is needed for exposed low level functions, and if any changes are needed to existing tests of fancy indexing. As suggested, I tried running `tox -e pre-commit` but got an error immediately due to distutils being removed. Running the tests, all passed except test_file.TestDrivers, which was not caused by this change.

iguinn committed:
Added combine_hyperslab, combine_select, modify_select and select_copy to h5s

iguinn committed:
Updated fancy indexing to use merge algorithm

iguinn committed:
Added release notes

neutrinoceros commented:
Not that I know of ! Sorry, I was just misremembering, it seems !

takluyver commented:
Thanks, nice work! I hadn't spotted when I opened #2536 that there was a way for h5py to get round the performance issue.

I've not looked through the changes in detail yet, but please do have a look through the tests and see if there's already a case that covers the new code paths or if we need to add one - I'm also not sure. I'd start with files like `test_dataset` and `test_dataset_getitem`, but indexing logic is a major part of h5py, so it could be in other modules.

takluyver commented on h5py/_selector.pyx:
```diff
@@ -258,10 +276,9 @@ cdef class Selector:
-            for i in array_arg:
-                tmp_start[array_ix] = i
-                H5Sselect_hyperslab(self.space, H5S_SELECT_OR, tmp_start, self.stride, tmp_count, self.block)
+            new_space = self.build_fancy_hyperslab(array_arg, array_ix, tmp_start, tmp_count)
+            H5Sselect_copy(self.space, new_space)
```
I'm not totally sure about this, but would it make sense for `build_fancy_hyperslab` to take space as a parameter and modify it, rather than always creating a new space and then copying the selection back to the original. For a large number of indices it wouldn't make much difference, because you'd have to create most of the spaces anyway, but it would presumably be a bit less overhead for smaller cases. :shrug: 

The divide-and-conquer part would then look something like this:

```cython
space2 = H5Scopy(space)
H5Sselect_none(space2)
self.build_fancy_hyperslab(space,  array_arg[:len(array_arg)//2], array_ix, tmp_start, tmp_count)
self.build_fancy_hyperslab(space2, array_arg[len(array_arg)//2:], array_ix, tmp_start, tmp_count)
H5Smodify_select(space, H5S_SELECT_OR, space2)
H5Sclose(space2)

iguinn replied:
I did this with one small difference. Instead of calling H5Scopy, which takes a noticeable amount of time to copy the selections existing in space (and then gets rid of it with select_none), it creates a new space.

takluyver reviewed (commented):
Thanks, this is mostly looking good. I have a few specific comments inline.

I would also like basic tests for the added low-level methods, as well as making sure we've got a fancy indexing test that uses the new code pathway. One way to check the latter is to intentionally raise an error in that branch and see if the tests fail.

iguinn committed:
Update h5py/api_functions.txt

Co-authored-by: Thomas Kluyver <takowl@gmail.com>

iguinn committed:
Update h5py/h5s.pyx

Co-authored-by: Thomas Kluyver <takowl@gmail.com>

iguinn committed:
Remove unnecessary copies when fancy indexing

iguinn committed:
Test indexlist with long enough list to trigger new logic

iguinn committed:
Change name of test

iguinn committed:
Fixed bug in h5s.select_copy

iguinn committed:
Added tests for h5s select_copy, combine_select, and modify_select

iguinn committed:
Removed h5s.combine_hyperslab

iguinn commented:
I've added tests now. When testing `h5s.combine_hyperslab` I found that it seems to be buggy; it seemed to work as expected only some of the time. This was with hdf5 v1.14.3. So I removed this function, which isn't important to the overall PR, and seems to be of limited utility since it is expected to be equivalent to `copy` followed by `select_hyperslab`. For the other functions, I only added very simple tests for the other low level functions for their basic functionality.

takluyver commented on h5py/tests/test_h5s.py:
```diff
@@ -23,3 +23,35 @@ def test_same_shape():
+    s1.select_hyperslab((5,), (1,), block=(5,), op=h5s.SELECT_SET)
+    s2 = h5s.create_simple((50,))
+    s2.select_copy(s1)
+    assert (s1.get_select_hyper_blocklist() == s2.get_select_hyper_blocklist()).all()
```
```suggestion
    np.testing.assert_array_equal(s1.get_select_hyper_blocklist(), s2.get_select_hyper_blocklist())
```

For numpy arrays, this can give more informative error messages if the assertion fails. (Same for the other 2 tests)

takluyver reviewed (commented):
Thanks @iguinn ! I think this is small enough to land in one go, we don't need to break it down further, I just haven't got round to looking at it for a long time. Now that I have done so, I'm pretty happy with it, I just have a few more minor comments.

takluyver committed:
Remove unused function from wrapper definitions

takluyver committed:
Refine docstrings for new methods

takluyver committed:
Polish new h5s tests

takluyver commented:
Thanks @iguinn !

gipert commented:
Hi, may I ask what's the timeline for releasing a new version that includes this patch?
